// Zoho_Scripts/zoho-doula-bio-system.js
/**
 * PHASE 4: DOULA BIO DELIVERY SYSTEM
 * Priya Sharma - Zoho Stack Specialist
 * Intelligent doula matching and profile delivery system
 */

const catalyst = require('zcatalyst-sdk-node');

class DoulaBioDeliverySystem {
    constructor() {
        this.app = catalyst.initialize();
        this.crm = this.app.crm();
        this.mail = this.app.sendmail();
        this.datastore = this.app.datastore();
        this.filestore = this.app.filestore();
        this.functions = this.app.functions();
    }

    /**
     * Initialize Doula Database with Sample Profiles
     */
    async initializeDoulaDatabase() {
        const doulaProfiles = [
            {
                doula_id: "DL001",
                first_name: "Sarah",
                last_name: "Martinez",
                certifications: ["DONA International", "CAPPA"],
                years_experience: 8,
                specializations: ["Natural Birth", "VBAC Support", "Postpartum Depression"],
                languages: ["English", "Spanish"],
                availability_zones: ["Downtown", "Westside", "North County"],
                birth_philosophy: "Evidence-based care with emotional support focus",
                rate_range: "$1200-1800",
                insurance_accepted: ["Aetna", "Blue Cross", "United Healthcare"],
                personality_type: "Calm and Nurturing",
                client_testimonial: "Sarah was incredible during my VBAC. Her knowledge and support made all the difference.",
                profile_image_url: "/images/doulas/sarah-martinez.jpg",
                contact_preference: "Text messaging preferred",
                emergency_availability: true,
                postpartum_services: true,
                bio_text: "Passionate about empowering families through evidence-based birth support. Specializes in high-risk pregnancies and VBAC preparation.",
                matching_keywords: ["natural", "vbac", "high-risk", "spanish", "nurturing", "experienced"]
            },
            {
                doula_id: "DL002", 
                first_name: "Jennifer",
                last_name: "Chen",
                certifications: ["Birth Boot Camp", "Spinning Babies"],
                years_experience: 5,
                specializations: ["Water Birth", "Hypnobirthing", "Breech Positioning"],
                languages: ["English", "Mandarin"],
                availability_zones: ["Eastside", "Central", "Airport Area"],
                birth_philosophy: "Gentle, intuitive birth support with focus on positioning",
                rate_range: "$1000-1500",
                insurance_accepted: ["Kaiser", "Cigna", "Anthem"],
                personality_type: "Gentle and Intuitive",
                client_testimonial: "Jennifer's calm presence and expertise with positioning techniques helped me achieve my dream water birth.",
                profile_image_url: "/images/doulas/jennifer-chen.jpg",
                contact_preference: "Email and video calls",
                emergency_availability: false,
                postpartum_services: true,
                bio_text: "Specializes in natural positioning techniques and creating serene birth environments. Certified in multiple comfort measure techniques.",
                matching_keywords: ["water", "hypnobirth", "gentle", "positioning", "mandarin", "natural"]
            },
            {
                doula_id: "DL003",
                first_name: "Maria",
                last_name: "Rodriguez",
                certifications: ["DONA International", "Childbirth International"],
                years_experience: 12,
                specializations: ["C-Section Support", "Twin/Multiple Births", "Premature Birth"],
                languages: ["English", "Spanish", "Portuguese"],
                availability_zones: ["South County", "Border Region", "Downtown"],
                birth_philosophy: "Comprehensive support for complex births with family-centered care",
                rate_range: "$1500-2200",
                insurance_accepted: ["All Major Insurance", "Medicaid"],
                personality_type: "Experienced and Reassuring",
                client_testimonial: "Maria's experience with twins was invaluable. She helped us navigate a complex delivery with confidence.",
                profile_image_url: "/images/doulas/maria-rodriguez.jpg",
                contact_preference: "Phone calls preferred",
                emergency_availability: true,
                postpartum_services: true,
                bio_text: "Veteran doula specializing in high-risk and multiple births. Bilingual support with extensive medical team collaboration experience.",
                matching_keywords: ["csection", "twins", "multiples", "high-risk", "experienced", "spanish", "portuguese", "medical"]
            },
            {
                doula_id: "DL004",
                first_name: "Ashley",
                last_name: "Thompson",
                certifications: ["CAPPA", "Postpartum Support International"],
                years_experience: 6,
                specializations: ["First-Time Parents", "Breastfeeding Support", "Mental Health"],
                languages: ["English"],
                availability_zones: ["Suburbs", "North County", "Westside"],
                birth_philosophy: "Education-focused support with emphasis on informed choices",
                rate_range: "$1100-1600",
                insurance_accepted: ["Blue Cross", "United", "Humana"],
                personality_type: "Educational and Supportive",
                client_testimonial: "Ashley made us feel so prepared and confident as first-time parents. Her knowledge was incredible.",
                profile_image_url: "/images/doulas/ashley-thompson.jpg",
                contact_preference: "Text and email combination",
                emergency_availability: true,
                postpartum_services: true,
                bio_text: "Dedicated to educating and empowering first-time parents. Strong background in breastfeeding support and postpartum mental health.",
                matching_keywords: ["first-time", "education", "breastfeeding", "mental-health", "supportive", "preparation"]
            }
        ];

        const doulaTable = this.datastore.table('doula_profiles');
        
        for (const profile of doulaProfiles) {
            try {
                await doulaTable.insertRow(profile);
                console.log(`✅ Doula profile created: ${profile.first_name} ${profile.last_name}`);
            } catch (error) {
                console.log(`📝 Profile exists: ${profile.doula_id}`);
            }
        }
        
        return doulaProfiles;
    }

    /**
     * Intelligent Doula Matching Algorithm
     */
    async matchDoulaProfiles(clientData) {
        const doulaTable = this.datastore.table('doula_profiles');
        const allDoulas = await doulaTable.getAllRows();
        
        // Extract client preferences from interview data
        const clientPreferences = this.parseClientPreferences(clientData);
        const matchedDoulas = [];

        for (const doula of allDoulas) {
            let matchScore = 0;
            const matchReasons = [];

            // Language matching (high priority)
            if (clientPreferences.preferred_language && 
                doula.languages.some(lang => 
                    lang.toLowerCase().includes(clientPreferences.preferred_language.toLowerCase())
                )) {
                matchScore += 25;
                matchReasons.push(`Speaks ${clientPreferences.preferred_language}`);
            }

            // Specialization matching
            if (clientPreferences.birth_type) {
                const birthKeywords = clientPreferences.birth_type.toLowerCase().split(' ');
                for (const keyword of birthKeywords) {
                    if (doula.matching_keywords.some(k => k.includes(keyword))) {
                        matchScore += 20;
                        matchReasons.push(`Specializes in ${clientPreferences.birth_type}`);
                        break;
                    }
                }
            }

            // Insurance matching
            if (clientPreferences.insurance_provider && 
                doula.insurance_accepted.some(ins => 
                    ins.toLowerCase().includes(clientPreferences.insurance_provider.toLowerCase())
                )) {
                matchScore += 15;
                matchReasons.push(`Accepts ${clientPreferences.insurance_provider}`);
            }

            // Geographic proximity
            if (clientPreferences.location_preference && 
                doula.availability_zones.some(zone => 
                    zone.toLowerCase().includes(clientPreferences.location_preference.toLowerCase())
                )) {
                matchScore += 10;
                matchReasons.push(`Available in ${clientPreferences.location_preference} area`);
            }

            // Experience level matching
            if (clientPreferences.experience_preference === 'experienced' && doula.years_experience >= 8) {
                matchScore += 15;
                matchReasons.push('Highly experienced (8+ years)');
            } else if (clientPreferences.experience_preference === 'newer' && doula.years_experience <= 5) {
                matchScore += 10;
                matchReasons.push('Newer doula with fresh energy');
            }

            // Emergency availability
            if (clientPreferences.high_risk && doula.emergency_availability) {
                matchScore += 12;
                matchReasons.push('Available for emergency situations');
            }

            // Personality matching
            if (clientPreferences.personality_preference && 
                doula.personality_type.toLowerCase().includes(clientPreferences.personality_preference.toLowerCase())) {
                matchScore += 8;
                matchReasons.push(`${doula.personality_type} personality style`);
            }

            // Add to matched list if score is above threshold
            if (matchScore >= 20) {
                matchedDoulas.push({
                    ...doula,
                    match_score: matchScore,
                    match_reasons: matchReasons,
                    compatibility_percentage: Math.min(Math.round((matchScore / 100) * 100), 98)
                });
            }
        }

        // Sort by match score and return top 3-4 matches
        return matchedDoulas
            .sort((a, b) => b.match_score - a.match_score)
            .slice(0, 4);
    }

    /**
     * Parse client preferences from CRM data
     */
    parseClientPreferences(clientData) {
        return {
            preferred_language: this.extractLanguagePreference(clientData.birth_preferences),
            birth_type: this.extractBirthTypePreference(clientData.birth_preferences),
            insurance_provider: clientData.insurance_provider,
            location_preference: this.extractLocationPreference(clientData),
            experience_preference: this.extractExperiencePreference(clientData.birth_preferences),
            high_risk: this.assessHighRiskFactors(clientData),
            personality_preference: this.extractPersonalityPreference(clientData.birth_preferences)
        };
    }

    extractLanguagePreference(birthPreferences) {
        if (!birthPreferences) return 'English';
        const text = birthPreferences.toLowerCase();
        if (text.includes('spanish') || text.includes('español')) return 'Spanish';
        if (text.includes('mandarin') || text.includes('chinese')) return 'Mandarin';
        if (text.includes('portuguese')) return 'Portuguese';
        return 'English';
    }

    extractBirthTypePreference(birthPreferences) {
        if (!birthPreferences) return 'natural';
        const text = birthPreferences.toLowerCase();
        if (text.includes('vbac')) return 'VBAC';
        if (text.includes('c-section') || text.includes('cesarean')) return 'C-Section';
        if (text.includes('water birth')) return 'Water Birth';
        if (text.includes('natural') || text.includes('unmedicated')) return 'Natural Birth';
        if (text.includes('twins') || text.includes('multiple')) return 'Multiple Births';
        return 'natural';
    }

    extractLocationPreference(clientData) {
        // This would typically come from client address or preferences
        return 'Downtown'; // Default, would parse from actual location data
    }

    extractExperiencePreference(birthPreferences) {
        if (!birthPreferences) return 'experienced';
        const text = birthPreferences.toLowerCase();
        if (text.includes('experienced') || text.includes('veteran')) return 'experienced';
        if (text.includes('newer') || text.includes('young')) return 'newer';
        return 'experienced';
    }

    assessHighRiskFactors(clientData) {
        if (!clientData.birth_preferences) return false;
        const text = clientData.birth_preferences.toLowerCase();
        return text.includes('high-risk') || text.includes('complications') || 
               text.includes('diabetes') || text.includes('preeclampsia') ||
               text.includes('breech') || text.includes('premature');
    }

    extractPersonalityPreference(birthPreferences) {
        if (!birthPreferences) return 'calm';
        const text = birthPreferences.toLowerCase();
        if (text.includes('calm') || text.includes('peaceful')) return 'calm';
        if (text.includes('energetic') || text.includes('encouraging')) return 'energetic';
        if (text.includes('gentle') || text.includes('soft')) return 'gentle';
        return 'nurturing';
    }

    /**
     * Generate Personalized Doula Bio Email
     */
    async generateDoulaBioEmail(clientData, matchedDoulas) {
        const doulaCardsHtml = matchedDoulas.map(doula => `
            <div style="border: 2px solid #e0e0e0; border-radius: 12px; padding: 20px; margin: 20px 0; background: white;">
                <div style="display: flex; align-items: flex-start; gap: 20px;">
                    <div style="min-width: 120px;">
                        <img src="${doula.profile_image_url}" alt="${doula.first_name}" 
                             style="width: 120px; height: 120px; border-radius: 50%; object-fit: cover; border: 3px solid #f0f0f0;">
                    </div>
                    <div style="flex: 1;">
                        <h3 style="color: #8b4513; margin: 0 0 10px 0; font-size: 22px;">
                            ${doula.first_name} ${doula.last_name}
                        </h3>
                        <div style="background: #e8f5e8; padding: 8px 12px; border-radius: 20px; display: inline-block; margin-bottom: 10px;">
                            <strong style="color: #2d5016;">${doula.compatibility_percentage}% Match</strong>
                        </div>
                        
                        <p style="color: #666; margin: 10px 0; line-height: 1.5;">
                            <strong>Experience:</strong> ${doula.years_experience} years<br>
                            <strong>Specializations:</strong> ${doula.specializations.join(', ')}<br>
                            <strong>Languages:</strong> ${doula.languages.join(', ')}
                        </p>
                        
                        <div style="background: #f8f4f1; padding: 12px; border-radius: 8px; margin: 10px 0;">
                            <p style="margin: 0; font-style: italic; color: #8b4513;">
                                "${doula.bio_text}"
                            </p>
                        </div>
                        
                        <div style="background: #e3f2fd; padding: 10px; border-radius: 6px; margin: 10px 0;">
                            <strong style="color: #1976d2;">Why we matched you:</strong>
                            <ul style="margin: 5px 0; padding-left: 20px;">
                                ${doula.match_reasons.map(reason => `<li style="color: #666;">${reason}</li>`).join('')}
                            </ul>
                        </div>
                        
                        <div style="background: #fff3e0; padding: 10px; border-radius: 6px; margin: 10px 0;">
                            <p style="margin: 0; color: #e65100;">
                                <strong>Rate Range:</strong> ${doula.rate_range} | 
                                <strong>Insurance:</strong> ${doula.insurance_accepted.join(', ')}
                            </p>
                        </div>
                        
                        <blockquote style="border-left: 4px solid #8b4513; padding-left: 15px; margin: 15px 0; font-style: italic; color: #666;">
                            "${doula.client_testimonial}"
                        </blockquote>
                    </div>
                </div>
            </div>
        `).join('');

        return `
        <div style="font-family: Arial, sans-serif; max-width: 700px; margin: 0 auto; background: #fafafa; padding: 20px;">
            <header style="background: linear-gradient(135deg, #8b4513, #d2691e); padding: 30px; text-align: center; border-radius: 12px; color: white;">
                <h1 style="margin: 0; font-size: 28px;">Meet Your Matched Doulas! 👥</h1>
                <p style="margin: 10px 0 0 0; font-size: 16px; opacity: 0.9;">Carefully selected based on your preferences</p>
            </header>
            
            <main style="padding: 20px 0;">
                <div style="background: white; padding: 25px; border-radius: 12px; margin-bottom: 20px; border: 1px solid #e0e0e0;">
                    <h2 style="color: #8b4513; margin-top: 0;">Hi ${clientData.first_name}! 🌟</h2>
                    
                    <p style="line-height: 1.6; color: #333;">
                        Based on our wonderful interview and your specific preferences, we've carefully selected these doulas who we believe would be perfect matches for your birth journey. Each has been chosen based on your unique needs and preferences.
                    </p>
                    
                    <div style="background: #e8f5e8; padding: 15px; border-radius: 8px; margin: 20px 0;">
                        <h3 style="margin-top: 0; color: #2d5016;">📋 Your Matching Criteria:</h3>
                        <ul style="margin: 10px 0; line-height: 1.6;">
                            <li>Due Date: Your upcoming delivery timeline</li>
                            <li>Birth Preferences: Supportive, personalized care</li>
                            <li>Insurance: Coverage verified and accepted</li>
                        </ul>
                    </div>
                </div>
                
                <h2 style="text-align: center; color: #8b4513; margin: 30px 0;">Your Personalized Doula Matches</h2>
                
                ${doulaCardsHtml}
                
                <div style="background: #f0f8ff; padding: 25px; border-radius: 12px; margin: 30px 0; border: 1px solid #e3f2fd;">
                    <h3 style="color: #2c5aa0; margin-top: 0;">📅 Ready for the Next Step?</h3>
                    <p style="line-height: 1.6; color: #333;">
                        We recommend scheduling brief meet-and-greet calls with 2-3 of these doulas to see who feels like the best personal fit. Most clients know within the first few minutes of conversation!
                    </p>
                    
                    <div style="text-align: center; margin: 20px 0;">
                        <a href="mailto:hello@snugandkisses.com?subject=Ready to Schedule Doula Meet-and-Greets" 
                           style="background: #2c5aa0; color: white; padding: 15px 30px; text-decoration: none; border-radius: 25px; display: inline-block; font-weight: bold;">
                            Schedule Meet-and-Greets 📞
                        </a>
                    </div>
                </div>
                
                <div style="background: #fff3e0; padding: 20px; border-radius: 8px; margin: 20px 0;">
                    <h4 style="margin-top: 0; color: #e65100;">💡 Questions to Ask During Meet-and-Greets:</h4>
                    <ul style="line-height: 1.6; color: #333;">
                        <li>How do you support partners during labor?</li>
                        <li>What's your approach to pain management?</li>
                        <li>How do you work with medical teams?</li>
                        <li>What does postpartum support look like?</li>
                        <li>Can you share a recent birth story?</li>
                    </ul>
                </div>
                
                <p style="text-align: center; margin: 30px 0; color: #666;">
                    Questions? Call us at <strong>(555) 123-4567</strong> or reply to this email.<br>
                    We're here to support you every step of the way! 💝
                </p>
            </main>
            
            <footer style="text-align: center; padding: 20px; color: #666; border-top: 1px solid #e0e0e0;">
                <p style="margin: 0;">
                    <strong>Snug & Kisses</strong><br>
                    Exceptional Postpartum Care<br>
                    hello@snugandkisses.com | (555) 123-4567
                </p>
            </footer>
        </div>`;
    }

    /**
     * Send Doula Profiles to Client
     */
    async sendDoulaProfiles(crmRecordId, clientData) {
        try {
            // Get matched doulas
            const matchedDoulas = await this.matchDoulaProfiles(clientData);
            
            if (matchedDoulas.length === 0) {
                console.log("❌ No suitable doula matches found");
                return { error: "No matches found" };
            }

            // Generate personalized email
            const emailHtml = await this.generateDoulaBioEmail(clientData, matchedDoulas);
            
            // Send email
            const emailConfig = {
                to: [{ 
                    email: clientData.email, 
                    name: clientData.first_name 
                }],
                from: { 
                    email: "hello@snugandkisses.com", 
                    name: "Snug & Kisses Care Team" 
                },
                subject: `Meet your matched doulas! 👥 ${matchedDoulas.length} perfect matches inside`,
                html_body: emailHtml
            };

            const emailResponse = await this.mail.sendMail(emailConfig);
            
            // Log in CRM
            await this.logDoulaDelivery(crmRecordId, matchedDoulas);
            
            // Update client status to Phase 4 Complete
            await this.updateClientPhase(crmRecordId, "Phase_4_Complete");
            
            console.log(`✅ Doula profiles sent to ${clientData.email}`);
            console.log(`📊 ${matchedDoulas.length} doulas matched and delivered`);
            
            return {
                success: true,
                matches_sent: matchedDoulas.length,
                doula_ids: matchedDoulas.map(d => d.doula_id),
                email_id: emailResponse.message_id
            };
            
        } catch (error) {
            console.error("❌ Doula profile delivery failed:", error);
            throw error;
        }
    }

    /**
     * Log doula delivery activity in CRM
     */
    async logDoulaDelivery(crmRecordId, matchedDoulas) {
        const activityData = {
            module: "Activities",
            data: {
                "Subject": `Phase 4: Doula Profiles Delivered`,
                "Activity_Type": "Email",
                "Status": "Completed",
                "What_Id": crmRecordId,
                "Activity_DateTime": new Date().toISOString(),
                "Description": `Delivered ${matchedDoulas.length} matched doula profiles: ${matchedDoulas.map(d => d.first_name + ' ' + d.last_name).join(', ')}`
            }
        };
        
        await this.crm.insertRecords(activityData);
    }

    /**
     * Update client phase status
     */
    async updateClientPhase(crmRecordId, newPhase) {
        const updateData = {
            module: "Leads",
            id: crmRecordId,
            data: { 
                "Stage": newPhase,
                "Lead_Status": "Doula_Profiles_Delivered",
                "Phase_4_Complete_Date": new Date().toISOString()
            }
        };
        
        await this.crm.updateRecord(updateData);
    }
}

module.exports = DoulaBioDeliverySystem;

// Catalyst Function Handler
const doulaBioSystem = new DoulaBioDeliverySystem();

exports.handler = async (event, context) => {
    try {
        const { action, data } = JSON.parse(event.body || '{}');
        
        switch (action) {
            case 'initialize_doulas':
                return await doulaBioSystem.initializeDoulaDatabase();
                
            case 'match_doulas':
                return await doulaBioSystem.matchDoulaProfiles(data.client_data);
                
            case 'send_profiles':
                return await doulaBioSystem.sendDoulaProfiles(data.crm_record_id, data.client_data);
                
            case 'generate_email':
                const matches = await doulaBioSystem.matchDoulaProfiles(data.client_data);
                return await doulaBioSystem.generateDoulaBioEmail(data.client_data, matches);
                
            default:
                return { 
                    statusCode: 400, 
                    body: JSON.stringify({ error: 'Invalid action' }) 
                };
        }
    } catch (error) {
        console.error("❌ Doula Bio System Error:", error);
        return { 
            statusCode: 500, 
            body: JSON.stringify({ error: error.message }) 
        };
    }
};
