// Zoho_Scripts/zoho-selection-process.js
/**
 * PHASE 6: CLIENT SELECTION PROCESS AUTOMATION
 * Priya Sharma - Zoho Stack Specialist
 * Final doula selection and contract initiation system
 */

const catalyst = require('zcatalyst-sdk-node');

class ClientSelectionProcess {
    constructor() {
        this.app = catalyst.initialize();
        this.crm = this.app.crm();
        this.mail = this.app.sendmail();
        this.datastore = this.app.datastore();
        this.filestore = this.app.filestore();
        this.functions = this.app.functions();
    }

    /**
     * Initialize Selection Process Components
     */
    async initializeSelectionProcess() {
        // Create selection tracking table
        const selectionTableConfig = {
            table_name: 'doula_selections',
            columns: [
                { column_name: 'client_id', data_type: 'varchar' },
                { column_name: 'selected_doula_id', data_type: 'varchar' },
                { column_name: 'runner_up_doula_id', data_type: 'varchar' },
                { column_name: 'selection_date', data_type: 'datetime' },
                { column_name: 'selection_reason', data_type: 'text' },
                { column_name: 'contract_status', data_type: 'varchar' },
                { column_name: 'contract_sent_date', data_type: 'datetime' },
                { column_name: 'contract_signed_date', data_type: 'datetime' },
                { column_name: 'phase_6_status', data_type: 'varchar' }
            ]
        };

        try {
            const table = this.datastore.table('doula_selections');
            console.log("✅ Selection tracking initialized");
            return table;
        } catch (error) {
            console.log("📝 Selection table already exists");
            return this.datastore.table('doula_selections');
        }
    }

    /**
     * Send Selection Decision Email
     */
    async sendSelectionReminderEmail(clientData, meetGreetHistory) {
        const selectionEmailHtml = this.generateSelectionReminderEmail(clientData, meetGreetHistory);
        
        const emailConfig = {
            to: [{ 
                email: clientData.email, 
                name: clientData.first_name 
            }],
            from: { 
                email: "hello@snugandkisses.com", 
                name: "Snug & Kisses Care Team" 
            },
            subject: `Ready to choose your doula? 💝 Let's finalize your perfect match!`,
            html_body: selectionEmailHtml
        };

        try {
            const emailResponse = await this.mail.sendMail(emailConfig);
            console.log("✅ Selection reminder sent:", emailResponse.message_id);
            return emailResponse;
        } catch (error) {
            console.error("❌ Selection reminder failed:", error);
            throw error;
        }
    }

    /**
     * Generate Selection Reminder Email
     */
    generateSelectionReminderEmail(clientData, meetGreetHistory) {
        const doulaReviewsHtml = meetGreetHistory.map(meeting => `
            <div style="border: 1px solid #e0e0e0; border-radius: 8px; padding: 20px; margin: 15px 0; background: white;">
                <div style="display: flex; align-items: center; gap: 15px; margin-bottom: 15px;">
                    <img src="${meeting.doula.profile_image_url}" alt="${meeting.doula.first_name}" 
                         style="width: 70px; height: 70px; border-radius: 50%; object-fit: cover;">
                    <div>
                        <h3 style="margin: 0; color: #8b4513;">${meeting.doula.first_name} ${meeting.doula.last_name}</h3>
                        <p style="margin: 5px 0; color: #666;">
                            ${meeting.doula.years_experience} years • ${meeting.doula.specializations.slice(0, 2).join(', ')}
                        </p>
                        <div style="background: #e8f5e8; padding: 4px 8px; border-radius: 12px; display: inline-block;">
                            <small style="color: #2d5016; font-weight: bold;">${meeting.doula.compatibility_percentage}% Match</small>
                        </div>
                    </div>
                </div>
                
                <div style="background: #f8f4f1; padding: 12px; border-radius: 6px; margin: 10px 0;">
                    <p style="margin: 0; font-style: italic; color: #8b4513; font-size: 14px;">
                        "${meeting.doula.bio_text}"
                    </p>
                </div>
                
                <div style="background: #fff3e0; padding: 12px; border-radius: 6px; margin: 10px 0;">
                    <p style="margin: 0; color: #e65100; font-size: 14px;">
                        <strong>Rate:</strong> ${meeting.doula.rate_range} | 
                        <strong>Insurance:</strong> ${meeting.doula.insurance_accepted.includes('All Major') ? 'All Major Insurance' : meeting.doula.insurance_accepted.slice(0, 2).join(', ')}
                    </p>
                </div>
                
                <div style="text-align: center; margin-top: 15px;">
                    <a href="mailto:hello@snugandkisses.com?subject=I choose ${meeting.doula.first_name} ${meeting.doula.last_name}&body=Hi! I'd like to select ${meeting.doula.first_name} ${meeting.doula.last_name} as my doula. Please send me the contract and next steps.%0D%0A%0D%0AClient: ${clientData.first_name} ${clientData.last_name}%0D%0ADoula ID: ${meeting.doula.doula_id}" 
                       style="background: #8b4513; color: white; padding: 10px 20px; text-decoration: none; border-radius: 25px; font-weight: bold; display: inline-block;">
                        Choose ${meeting.doula.first_name} 💝
                    </a>
                </div>
            </div>
        `).join('');

        return `
        <div style="font-family: Arial, sans-serif; max-width: 650px; margin: 0 auto; background: #fafafa; padding: 20px;">
            <header style="background: linear-gradient(135deg, #8b4513, #d2691e); padding: 25px; text-align: center; border-radius: 12px; color: white;">
                <h1 style="margin: 0; font-size: 28px;">Ready to Choose Your Doula? 💝</h1>
                <p style="margin: 10px 0 0 0; opacity: 0.9;">Let's finalize your perfect birth support partner</p>
            </header>
            
            <main style="padding: 20px 0;">
                <div style="background: white; padding: 25px; border-radius: 12px; margin-bottom: 20px;">
                    <h2 style="color: #8b4513; margin-top: 0;">Hi ${clientData.first_name}! 🌟</h2>
                    
                    <p style="line-height: 1.6; color: #333;">
                        You've had some wonderful conversations with our doulas! Now comes the exciting part - 
                        choosing your birth support partner. Trust your instincts - you'll know who felt like the right fit.
                    </p>
                    
                    <div style="background: #e3f2fd; padding: 15px; border-radius: 8px; margin: 20px 0;">
                        <h4 style="margin-top: 0; color: #1976d2;">💭 Things to Consider:</h4>
                        <ul style="margin: 5px 0; line-height: 1.5; color: #333;">
                            <li>Who made you feel most comfortable and understood?</li>
                            <li>Whose birth philosophy aligned best with your vision?</li>
                            <li>Who do you feel your partner connected with?</li>
                            <li>Who seemed most experienced with your specific needs?</li>
                        </ul>
                    </div>
                </div>
                
                <h3 style="color: #8b4513; text-align: center; margin: 25px 0;">Your Doula Options</h3>
                
                ${doulaReviewsHtml}
                
                <div style="background: #e8f5e8; padding: 25px; border-radius: 12px; margin: 25px 0; border: 1px solid #c8e6c9;">
                    <h3 style="color: #2d5016; margin-top: 0; text-align: center;">🎯 Ready to Decide?</h3>
                    <p style="text-align: center; color: #333; margin-bottom: 20px;">
                        Simply click "Choose [Name]" above for your selected doula, or email us with your decision!
                    </p>
                    
                    <div style="text-align: center;">
                        <a href="mailto:hello@snugandkisses.com?subject=Need Help Choosing My Doula&body=Hi! I met with several wonderful doulas but need some guidance in making my final decision. Can we schedule a quick call?%0D%0A%0D%0AClient: ${clientData.first_name} ${clientData.last_name}" 
                           style="background: #2d5016; color: white; padding: 12px 25px; text-decoration: none; border-radius: 25px; display: inline-block; font-weight: bold; margin: 0 10px;">
                            Need Help Deciding? 🤔
                        </a>
                    </div>
                </div>
                
                <div style="background: #f0f8ff; padding: 20px; border-radius: 8px; margin: 20px 0; border-left: 4px solid #2c5aa0;">
                    <h4 style="margin-top: 0; color: #2c5aa0;">📋 What Happens Next:</h4>
                    <ol style="line-height: 1.6; color: #333; margin: 10px 0;">
                        <li><strong>You choose your doula</strong> - Click the button or email us</li>
                        <li><strong>We send your contract</strong> - Usually within 2-4 hours</li>
                        <li><strong>Review and sign</strong> - Digital signing, very simple</li>
                        <li><strong>Welcome call scheduled</strong> - Get to know your doula better</li>
                        <li><strong>Birth plan creation</strong> - Personalized support planning</li>
                    </ol>
                </div>
                
                <div style="background: #fff3e0; padding: 20px; border-radius: 8px; margin: 20px 0;">
                    <h4 style="margin-top: 0; color: #e65100;">⏰ Timeline Reminders:</h4>
                    <ul style="line-height: 1.5; color: #333;">
                        <li><strong>Your due date:</strong> ${clientData.due_date}</li>
                        <li><strong>Ideal decision time:</strong> Within next 2-3 days</li>
                        <li><strong>Contract processing:</strong> 24-48 hours</li>
                        <li><strong>Welcome call:</strong> Within 1 week of signing</li>
                    </ul>
                </div>
                
                <div style="text-align: center; margin: 30px 0; padding: 20px; background: white; border-radius: 12px; border: 2px solid #8b4513;">
                    <h3 style="color: #8b4513; margin-top: 0;">Still Have Questions? 💬</h3>
                    <p style="color: #666; margin-bottom: 20px;">
                        Our team is here to help! We can provide additional insights, 
                        answer specific questions, or help you feel confident in your choice.
                    </p>
                    <div>
                        <a href="tel:5551234567" 
                           style="background: white; color: #8b4513; border: 2px solid #8b4513; padding: 10px 20px; text-decoration: none; border-radius: 25px; display: inline-block; font-weight: bold; margin: 0 5px;">
                            Call (555) 123-4567 📞
                        </a>
                        <a href="mailto:hello@snugandkisses.com?subject=Questions About My Doula Selection" 
                           style="background: #8b4513; color: white; padding: 10px 20px; text-decoration: none; border-radius: 25px; display: inline-block; font-weight: bold; margin: 0 5px;">
                            Send Email 💌
                        </a>
                    </div>
                </div>
                
                <p style="text-align: center; margin: 25px 0; color: #666; font-style: italic;">
                    "The right doula will feel like a natural extension of your birth team. Trust that feeling!" 🌟
                </p>
            </main>
            
            <footer style="text-align: center; padding: 20px; color: #666; border-top: 1px solid #e0e0e0;">
                <p style="margin: 0;">
                    <strong>Snug & Kisses</strong> • Exceptional Postpartum Care<br>
                    hello@snugandkisses.com • (555) 123-4567
                </p>
            </footer>
        </div>`;
    }

    /**
     * Process Doula Selection
     */
    async processDoulaSelection(clientData, selectedDoulaId, selectionReason = '') {
        try {
            const selectionTable = this.datastore.table('doula_selections');
            const doulaTable = this.datastore.table('doula_profiles');
            
            // Get selected doula details
            const selectedDoula = await doulaTable.getRow(selectedDoulaId);
            
            // Record selection
            const selectionRecord = {
                client_id: clientData.crm_id,
                selected_doula_id: selectedDoulaId,
                selection_date: new Date().toISOString(),
                selection_reason: selectionReason,
                contract_status: 'pending',
                phase_6_status: 'selection_made'
            };
            
            await selectionTable.insertRow(selectionRecord);
            
            // Update CRM with selection
            await this.updateCRMWithSelection(clientData.crm_id, selectedDoula);
            
            // Send confirmation emails
            await this.sendSelectionConfirmations(clientData, selectedDoula);
            
            // Generate and send contract
            await this.initiateContractProcess(clientData, selectedDoula);
            
            // Notify selected doula
            await this.notifySelectedDoula(selectedDoula, clientData);
            
            console.log(`✅ Selection processed: ${clientData.first_name} chose ${selectedDoula.first_name}`);
            
            return {
                success: true,
                selected_doula: `${selectedDoula.first_name} ${selectedDoula.last_name}`,
                contract_status: 'initiated',
                next_steps: 'Contract sent for review and signing'
            };
            
        } catch (error) {
            console.error("❌ Selection processing failed:", error);
            throw error;
        }
    }

    /**
     * Update CRM with Selection
     */
    async updateCRMWithSelection(crmRecordId, selectedDoula) {
        const updateData = {
            module: "Leads",
            id: crmRecordId,
            data: {
                "Stage": "Phase_6_Complete",
                "Lead_Status": "Doula_Selected",
                "Selected_Doula": `${selectedDoula.first_name} ${selectedDoula.last_name}`,
                "Selected_Doula_ID": selectedDoula.doula_id,
                "Selection_Date": new Date().toISOString(),
                "Contract_Status": "Pending_Signature",
                "Next_Action": "Contract_Review"
            }
        };
        
        await this.crm.updateRecord(updateData);
        
        // Log activity
        const activityData = {
            module: "Activities",
            data: {
                "Subject": `PHASE 6 COMPLETE: Doula Selected - ${selectedDoula.first_name} ${selectedDoula.last_name}`,
                "Activity_Type": "Task",
                "Status": "Completed",
                "What_Id": crmRecordId,
                "Activity_DateTime": new Date().toISOString(),
                "Description": `Client completed Phase 6 selection process. Selected doula: ${selectedDoula.first_name} ${selectedDoula.last_name} (${selectedDoula.doula_id}). Contract initiated.`
            }
        };
        
        await this.crm.insertRecords(activityData);
    }

    /**
     * Send Selection Confirmation Emails
     */
    async sendSelectionConfirmations(clientData, selectedDoula) {
        // Client confirmation
        const clientConfirmationHtml = `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
            <header style="background: linear-gradient(135deg, #2d5016, #4caf50); padding: 25px; text-align: center; border-radius: 12px; color: white;">
                <h1 style="margin: 0; font-size: 26px;">Perfect Choice! 🎉</h1>
                <p style="margin: 10px 0 0 0; opacity: 0.9;">Your doula selection is confirmed</p>
            </header>
            
            <main style="padding: 25px 20px;">
                <h2 style="color: #2d5016;">Congratulations ${clientData.first_name}! 💝</h2>
                
                <p style="line-height: 1.6;">
                    Wonderful news! You've selected <strong>${selectedDoula.first_name} ${selectedDoula.last_name}</strong> as your doula. 
                    We're so excited about this match - ${selectedDoula.first_name} is thrilled to support your birth journey!
                </p>
                
                <div style="background: #e8f5e8; padding: 20px; border-radius: 8px; margin: 20px 0; border: 1px solid #c8e6c9;">
                    <h3 style="margin-top: 0; color: #2d5016;">📋 What's Happening Now:</h3>
                    <ul style="line-height: 1.6; color: #333;">
                        <li>✅ Your selection has been confirmed</li>
                        <li>📄 Contract being prepared (you'll receive it within 2-4 hours)</li>
                        <li>📞 Welcome call will be scheduled once contract is signed</li>
                        <li>📝 Birth plan creation will begin after welcome call</li>
                    </ul>
                </div>
                
                <div style="background: #f0f8ff; padding: 15px; border-radius: 8px; margin: 20px 0;">
                    <h4 style="margin-top: 0; color: #2c5aa0;">💌 About Your Doula:</h4>
                    <p style="margin: 0; color: #333;">
                        <strong>${selectedDoula.first_name}</strong> brings ${selectedDoula.years_experience} years of experience 
                        and specializes in ${selectedDoula.specializations.slice(0, 2).join(' and ')}. 
                        ${selectedDoula.first_name}'s approach: "${selectedDoula.birth_philosophy}"
                    </p>
                </div>
                
                <p style="text-align: center; margin: 25px 0;">
                    Keep an eye on your inbox for the contract, and don't hesitate to reach out with any questions! 🌟
                </p>
            </main>
        </div>`;

        const clientEmailConfig = {
            to: [{ 
                email: clientData.email, 
                name: clientData.first_name 
            }],
            from: { 
                email: "hello@snugandkisses.com", 
                name: "Snug & Kisses Care Team" 
            },
            subject: `Perfect choice! ${selectedDoula.first_name} is your doula! 🎉`,
            html_body: clientConfirmationHtml
        };

        await this.mail.sendMail(clientEmailConfig);
    }

    /**
     * Notify Selected Doula
     */
    async notifySelectedDoula(selectedDoula, clientData) {
        const doulaNotificationHtml = `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
            <header style="background: #8b4513; padding: 20px; text-align: center; border-radius: 12px; color: white;">
                <h1 style="margin: 0;">You've Been Selected! 🎉</h1>
                <p style="margin: 5px 0 0 0;">New client match confirmed</p>
            </header>
            
            <main style="padding: 25px 20px;">
                <h2 style="color: #8b4513;">Congratulations ${selectedDoula.first_name}! 💝</h2>
                
                <p style="line-height: 1.6;">
                    Wonderful news! <strong>${clientData.first_name} ${clientData.last_name}</strong> has chosen you as their doula. 
                    Your meet-and-greet clearly made a great impression!
                </p>
                
                <div style="background: #f8f4f1; padding: 20px; border-radius: 8px; margin: 20px 0;">
                    <h3 style="margin-top: 0; color: #8b4513;">👶 Client Details:</h3>
                    <p><strong>Name:</strong> ${clientData.first_name} ${clientData.last_name}</p>
                    <p><strong>Due Date:</strong> ${clientData.due_date}</p>
                    <p><strong>Email:</strong> ${clientData.email}</p>
                    <p><strong>Phone:</strong> ${clientData.phone}</p>
                    <p><strong>Special Preferences:</strong> ${clientData.birth_preferences || 'Standard care preferences'}</p>
                </div>
                
                <div style="background: #e3f2fd; padding: 15px; border-radius: 8px; margin: 20px 0;">
                    <h4 style="margin-top: 0; color: #1976d2;">📋 Next Steps:</h4>
                    <ol style="line-height: 1.6; color: #333;">
                        <li>Client will receive and sign contract (usually within 24-48 hours)</li>
                        <li>We'll coordinate your welcome call once contract is signed</li>
                        <li>Begin birth plan discussions during welcome call</li>
                        <li>Establish regular check-in schedule</li>
                    </ol>
                </div>
                
                <p style="text-align: center;">
                    We'll keep you updated on contract status. Thank you for being part of the Snug & Kisses team! 🌟
                </p>
            </main>
        </div>`;

        const doulaEmailConfig = {
            to: [{ 
                email: selectedDoula.email, 
                name: `${selectedDoula.first_name} ${selectedDoula.last_name}` 
            }],
            from: { 
                email: "team@snugandkisses.com", 
                name: "Snug & Kisses Team" 
            },
            subject: `Congratulations! You've been selected by ${clientData.first_name}! 🎉`,
            html_body: doulaNotificationHtml
        };

        await this.mail.sendMail(doulaEmailConfig);
    }

    /**
     * Initiate Contract Process (Phase 7 preparation)
     */
    async initiateContractProcess(clientData, selectedDoula) {
        // This would integrate with contract generation system
        // For now, we'll create a placeholder that triggers Phase 7
        
        const contractData = {
            client_id: clientData.crm_id,
            doula_id: selectedDoula.doula_id,
            contract_type: 'standard_doula_agreement',
            rate: selectedDoula.rate_range,
            services_included: [
                'Prenatal consultation',
                'Labor and delivery support',
                'Postpartum check-in',
                'Birth plan development',
                '24/7 on-call availability'
            ],
            generated_date: new Date().toISOString(),
            status: 'pending_client_review'
        };

        // Store contract data for Phase 7 processing
        const contractTable = this.datastore.table('contracts');
        await contractTable.insertRow(contractData);
        
        console.log("📄 Contract process initiated - ready for Phase 7");
        
        return contractData;
    }

    /**
     * Generate Phase Completion Summary
     */
    async generatePhaseCompletionSummary(clientData) {
        const summaryHtml = `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
            <header style="background: linear-gradient(135deg, #8b4513, #d2691e); padding: 25px; text-align: center; border-radius: 12px; color: white;">
                <h1 style="margin: 0;">Phases 2-6 Complete! 🎯</h1>
                <p style="margin: 10px 0 0 0;">Your doula journey progress</p>
            </header>
            
            <main style="padding: 25px 20px;">
                <h2 style="color: #8b4513;">Amazing Progress ${clientData.first_name}! 🌟</h2>
                
                <div style="background: #e8f5e8; padding: 20px; border-radius: 8px; margin: 20px 0;">
                    <h3 style="margin-top: 0; color: #2d5016;">✅ What You've Accomplished:</h3>
                    <ul style="line-height: 1.8; color: #333;">
                        <li><strong>Phase 2:</strong> Scheduled and completed your initial interview</li>
                        <li><strong>Phase 3:</strong> Received personalized email guidance</li>
                        <li><strong>Phase 4:</strong> Got matched with perfect doula candidates</li>
                        <li><strong>Phase 5:</strong> Conducted meet-and-greet calls</li>
                        <li><strong>Phase 6:</strong> Selected your perfect doula match!</li>
                    </ul>
                </div>
                
                <div style="background: #f0f8ff; padding: 20px; border-radius: 8px; margin: 20px 0;">
                    <h3 style="margin-top: 0; color: #1976d2;">🚀 Coming Up Next:</h3>
                    <p style="margin: 0; line-height: 1.6;">
                        <strong>Phase 7-12:</strong> Contract signing, welcome calls, birth plan development, 
                        and comprehensive preparation for your birth experience. Your doula will be your 
                        dedicated partner through the rest of your journey!
                    </p>
                </div>
                
                <p style="text-align: center; margin: 25px 0; font-size: 18px; color: #8b4513;">
                    <strong>You're in amazing hands! 💝</strong>
                </p>
            </main>
        </div>`;

        return summaryHtml;
    }
}

module.exports = ClientSelectionProcess;

// Catalyst Function Handler
const selectionProcess = new ClientSelectionProcess();

exports.handler = async (event, context) => {
    try {
        const { action, data } = JSON.parse(event.body || '{}');
        
        switch (action) {
            case 'initialize_selection':
                return await selectionProcess.initializeSelectionProcess();
                
            case 'send_selection_reminder':
                return await selectionProcess.sendSelectionReminderEmail(
                    data.client_data, 
                    data.meetgreet_history
                );
                
            case 'process_selection':
                return await selectionProcess.processDoulaSelection(
                    data.client_data, 
                    data.selected_doula_id, 
                    data.selection_reason
                );
                
            case 'generate_summary':
                return await selectionProcess.generatePhaseCompletionSummary(data.client_data);
                
            default:
                return { 
                    statusCode: 400, 
                    body: JSON.stringify({ error: 'Invalid action' }) 
                };
        }
    } catch (error) {
        console.error("❌ Selection Process Error:", error);
        return { 
            statusCode: 500, 
            body: JSON.stringify({ error: error.message }) 
        };
    }
};
